File: game_plugins/hase_und_igel_2018/shared/sc/plugin2018/FallBack.java
package sc.plugin2018;

import com.thoughtworks.xstream.annotations.XStreamAlias;
import sc.plugin2018.util.GameRuleLogic;
import sc.shared.InvalidMoveException;

/**
 * Rückzugaktion. Sollte das nächste Igelfeld hinter einem Spieler nicht belegt sein, darf anstatt nach
 * vorne zu ziehen ein Rückzug gemacht werden. Dabei werden die zurückgezogene Distanz * 10 Karotten aufgenommen.
 */
@XStreamAlias(value = "fallBack")
public class FallBack extends Action {

  public FallBack() {
    this.order = 0;
  }

  public FallBack(int order) {
    this.order = order;
  }

  @Override
  public void perform(GameState state) throws InvalidMoveException {
    if (GameRuleLogic.isValidToFallBack(state)) {
      int previousFieldIndex = state.getCurrentPlayer().getFieldIndex();
      state.getCurrentPlayer().setFieldIndex(state.getPreviousFieldByType(FieldType.HEDGEHOG, state.getCurrentPlayer()
              .getFieldIndex()));
      state.getCurrentPlayer().changeCarrotsBy(10 * (previousFieldIndex - state.getCurrentPlayer().getFieldIndex()));
      state.setLastAction(this);
    } else {
      throw new InvalidMoveException("Es kann gerade kein Rückzug gemacht werden.");
    }
  }

  @Override
  public FallBack clone() {
    return new FallBack(this.order);
  }

  @Override
  public boolean equals(Object o) {
    if(o instanceof FallBack) {
      return true;
    }
    return false;
  }

  @Override
  public String toString() {
    return "FallBack order " + this.order;
  }
}

File: plugin/src/main/kotlin/sc/plugin2025/FallBack.kt
package sc.plugin2025

import com.thoughtworks.xstream.annotations.XStreamAlias

/**
 * Rückwärtszug.
 * Sollte das nächste Igelfeld hinter einem Spieler nicht belegt sein,
 * darf anstatt nach vorne zu ziehen ein Rückzug gemacht werden.
 * Dabei werden die zurückgezogene Distanz * 10 Karotten aufgenommen.
 */
@XStreamAlias(value = "fallBack")
object FallBack: HuIMove {
    override fun perform(state: GameState): MoveMistake? {
        val previousFieldIndex = state.nextFallBack()
        if(previousFieldIndex != null) {
            state.currentPlayer.carrots += 10 * (state.currentPlayer.position - previousFieldIndex)
            state.currentPlayer.position = previousFieldIndex
            return null
        } else {
            return MoveMistake.CANNOT_FALL_BACK
        }
    }
    
    override fun equals(other: Any?): Boolean = other is FallBack
}
